zerver/tests/test_channel_permissions.py

```diff
@@ -746,3 +748,229 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+                )
+                check_channel_property_update(hamlet)
+
+                # Reset the setting values to "Nobody" group.
+                do_change_stream_group_based_setting(
+                    stream, "can_add_subscribers_group", nobody_group, acting_user=iago
+                )
+                do_change_stream_group_based_setting(
+                    stream, "can_subscribe_group", nobody_group, acting_user=iago
+                )
+
+            self.subscribe(iago, stream.name)
+            self.subscribe(shiva, stream.name)
+            self.subscribe(hamlet, stream.name)
+
+    def test_channel_administering_permissions(self) -> None:
+        """
+        This test is only for checking channel administer permissions
+        while updating different channel properties. Other things
+        while updating these properties are tested separately.
+        """
+        realm = get_realm("zulip")
+        iago = self.example_user("iago")
+        public_stream = self.make_stream("test stream")
+
+        private_stream = self.make_stream("private_stream", invite_only=True)
+        self.subscribe(iago, private_stream.name)
+        self.subscribe(self.example_user("shiva"), private_stream.name)
+        self.subscribe(self.example_user("hamlet"), private_stream.name)
+        self.subscribe(self.example_user("prospero"), private_stream.name)
+
+        channel_folder = check_add_channel_folder(realm, "Frontend", "", acting_user=iago)
+        hamletcharacters_group = NamedUserGroup.objects.get(name="hamletcharacters", realm=realm)
+
+        for stream in [public_stream, private_stream]:
+            self.do_test_channel_administering_permission(stream, "name", "Renamed stream")
+            self.do_test_channel_administering_permission(
+                stream, "description", "Edited stream description"
+            )
+            self.do_test_channel_administering_permission(stream, "folder_id", channel_folder.id)
+
+            for setting_name in Stream.stream_permission_group_settings:
+                self.do_test_channel_administering_permission(
+                    stream, setting_name, hamletcharacters_group.id
+                )
+
+        moderators_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MODERATORS, realm=realm, is_system_group=True
+        )
+        do_change_realm_permission_group_setting(
+            realm, "can_create_web_public_channel_group", moderators_group, acting_user=None
+        )
+
+        # Test making a public stream private with protected history.
+        self.do_test_channel_administering_permission(public_stream, "invite_only", True)
+
+        # Test making a public stream web-public.
+        self.do_test_channel_administering_permission(public_stream, "is_web_public", True)
+
+        # Test making a private stream with protected history public.
+        self.do_test_channel_administering_permission(private_stream, "invite_only", False)
```

Comment:
Currently for changing stream privacy, this only tests cases that were checked before. We can probably cover other cases as well like changing private stream to web-public, etc.

Reply:
Sure, that does seem like a helpful follow-up.